Mathlib/GroupTheory/GroupAction/Primitive.lean

```diff
@@ -275,4 +290,113 @@ instance (priority := 100) IsPreprimitive.isQuasiPreprimitive [IsPreprimitive M
 
 end Normal
 
+section Finite
+
+namespace IsPreprimitive
+
+variable {M : Type*} [Group M] {α : Type*} [MulAction M α]
+variable {N β : Type*} [Group N] [MulAction N β]
+
+open scoped BigOperators Pointwise
+
+/-- A pretransitive action on a set of prime order is preprimitive -/
+@[to_additive "A pretransitive action on a set of prime order is preprimitive"]
+theorem of_prime_card [hGX : IsPretransitive M α] (hp : Nat.Prime (Nat.card α)) :
+    IsPreprimitive M α := by
+  refine ⟨fun {B} hB ↦ B.subsingleton_or_nontrivial.imp id fun hB' ↦ ?_⟩
+  have : Finite α := (Nat.card_ne_zero.mp hp.ne_zero).2
+  rw [Set.eq_univ_iff_ncard, eq_comm, ← hp.dvd_iff_eq ((Set.one_lt_ncard).mpr hB').ne']
+  exact hB.ncard_dvd_card hB'.nonempty
+
+variable {φ : M → N} {f : α →ₑ[φ] β}
+
+/-- The target of an equivariant map of large image is preprimitive if the source is -/
+@[to_additive "The target of an equivariant map of large image is preprimitive if the source is"]
+theorem of_card_lt [Finite β] [IsPretransitive N β] [IsPreprimitive M α]
+    (hf' : Nat.card β < 2 * (Set.range f).ncard) :
+    IsPreprimitive N β :=  by
+  classical
+  apply IsPreprimitive.mk
+  intro B hB
+  rcases B.eq_empty_or_nonempty with hB' | hB'
+  · left
+    rw [hB']
+    apply Set.subsingleton_empty
+  rw [IsTrivialBlock, or_iff_not_imp_right]
+  intro hB_ne_top
+  -- we need Set.Subsingleton B ↔ Set.ncard B ≤ 1
+  suffices Set.ncard B < 2 by
+    rw [Nat.lt_succ, Set.ncard_le_one_iff] at this
+    exact fun ⦃x⦄ x_1 ⦃y⦄ ↦ this x_1
```

Comment:
Our API should shield you from things that are this fiddly so it looks like we have a gap. Indeed, if we add the following just below `ncard_le_one` in `Data/Set/Card.lean`:
```lean
@[simp] theorem ncard_le_one_iff_subsingleton [Finite s] :
    s.ncard ≤ 1 ↔ s.Subsingleton :=
  ncard_le_one <| inferInstanceAs (Finite s)
```
then this can be just:
```suggestion
  suffices Set.ncard B < 2 by simpa [Nat.lt_succ] using this
```